src/witness/advancing_range.rs

```diff
@@ -0,0 +1,114 @@
+use std::ops::Range;
+
+/// Used for finding ranges in an array without having to iterate over the entire array every time.
+///
+/// It is assumed that the array contains u32 or tuples where the first element is u32. Let's call those
+/// number the keys.
+/// Provided that the keys are in ascending order, [AdvancingRange::get_slice] can be used to get a slice
+/// that contains all the elements with keys within a certain range.
+pub struct AdvancingRange<'a, T: TupleFirst> {
+    data: &'a [T],
+    start: usize,
+    end: usize,
+    previous_range: Range<u32>,
+}
+
+impl<'a, T: TupleFirst> AdvancingRange<'a, T> {
+    pub fn new(slice: &'a [T]) -> Self {
+        Self {
+            data: slice,
+            start: 0,
+            end: 0,
+            previous_range: 0..0,
+        }
+    }
+
+    /// Returns the range of the elements within `acceptable_range`, assuming
+    /// the elements are in nondecreasing order.
+    /// # Panics
+    /// Panics if `acceptable_range`'s start or end is less than in the previous call.
+    pub fn get_range(&mut self, acceptable_range: Range<u32>) -> Range<usize> {
+        assert!(acceptable_range.start >= self.previous_range.start);
+        assert!(acceptable_range.end >= self.previous_range.end);
+        self.previous_range = acceptable_range.clone();
+
+        for x in self.data[self.start..].iter() {
```

Comment:
potential optimisation (but not sure how often this happens):

if self.previous_range.end == acceptable_range.start --> then we could 'immediately' jump self.start to self.end, right?

Reply:
We could do that but this code is already very fast, so I don't think it will improve overall performance measurably.